### ISSUE TITLE
Serialization Problem

### ISSUE THREAD

**Issue by michidk** [wojtpl2/ExtendedXmlSerializer#26]
Hi,

I want to deserialize a xml file into a class (Scene.cs) containing this variable:
`        public List<GameObject> SceneGraph { get; set; }` 
This is the GameObject class: I'm trying to serialize this class:
https://github.com/michidk/BrokenEngine/blob/master/BrokenEngine/GameObject.cs

My xml file currently looks like this:
```
<?xml version="1.0"?>
<Scene xmlns="BrokenEngine.Scene">
  <SceneGraph>
    <GameObject type="BrokenEngine.GameObject">
      <Name>Super Cool GameObject</Name>
      <Position type="OpenTK.Vector3">
        <X>1</X>
        <Y>1</Y>
        <Z>1</Z>
      </Position>
    </GameObject>
  </SceneGraph>
</Scene>
```

I've already added a public, parameterless constructor.
I get the following error: {"Value cannot be null.\r\nParameter name: method"}
What do I do wrong?

**Comment by wojtpl2** [comment]
I'm not sure that version 1.5.0 support classes that inherit from IEnumerable. I have to check it.

**Comment by michidk** [comment]
Is there another version where this feature is supported? Or do I have to switch back to the default C# XMLSerializer?

**Comment by Mike-EEE** [comment]
Hey @michidk we are currently working on v2, which can be found here:
https://github.com/wojtpl2/ExtendedXmlSerializer/tree/v2.0.0

However, I am unsure how helpful that will be for you.  While it does address some problems, it is rather slow and only has a serializer in it.  The deserializer is from 1.0 and doesn't support all the new stuff we're (hoping 😄) to put into it.  If all goes well we should have a RC sometime late next week, I am thinking at this point. @wojtpl2 is tied up with his work but I have managed to find myself with free time this holiday season.  Lucky me. :)

**Comment by michidk** [comment]
Hey,
how is v2 coming along? 
Are there any news regarding this issue?

**Comment by Mike-EEE** [comment]
Hi @michidk thanks for checking in... as of right now I am anticipating another 2 weeks or so before being feature complete.  We have both serializer and deserializer working, and the performance is actually pretty decent (faster than the native `XmlSerializer` according to Benchmark.NET).

So for this issue, we currently would not emit your `GameObject` class as it does not inherit/implement `ICollection<>` or `IList`.  We could certainly serialize/emit the children (iterate through `IEnumerable`), but the problem would be reading the items back in (deserialize them).

It would be nice to hear your thoughts on this matter, or what your expectations are.

**Comment by michidk** [comment]
Great job so far!

It would be no problem for me to change GameObject so that it implements ICollection. Would this make this class de-/serializeable? If so, this is great and all I need.

**Comment by Mike-EEE** [comment]
Great!  Good to hear.  To be sure, `GameObject` is already serializable as long as it has a public constructor.  The one thing it will not do is not serialize the children unless it implements `IList` or `ICollection` as noted above.

To give some context, when EXS encounters each element in a graph, what it does is it uses [a selector to determine the type of element it is](https://github.com/wojtpl2/ExtendedXmlSerializer/blob/v2.0.0/src/ExtendedXmlSerializer/ConverterModel/Elements/ContainerDefinitions.cs#L43-L48), and it does so by going from specific to general.  So in this order:

- "Known" or registered types ([primitives/value types](https://github.com/wojtpl2/ExtendedXmlSerializer/tree/v2.0.0/src/ExtendedXmlSerializer/ConverterModel/Elements/Converters))
- Dictionary
- Array
- Collection (the condition we are talking here)
- "Membered" objects (what `GameObject` is as defined above -- without an `ICollection` implementation)
- General: catch-all, decided at runtime to be one of the above.  This is used for generalized/variable member types of object properties.

So if you try to serialize `GameObject` and it only implements `IEnumerable` (and not `ICollection`), it will serialize as a membered object, and serialize all of its fields and properties.  At least it _should_, without any exceptions. :)  It would be great to test your scenario once this is complete.

**Comment by michidk** [comment]
Gotcha. The children are GameObjects itsself, so GameObject hast to implement ICollection.
Also I have to make `private readonly List<GameObject> children` public, right?

**Comment by Mike-EEE** [comment]
Correct, @michidk.  In fact I believe at the moment if you place an `XmlElementAttribute` on the private `children` field it will pick it up as a member to emit. Since it is a `List<>`, it will satisfy the Collection condition above and emit accordingly.

**Comment by Mike-EEE** [comment]
Alright @michidk I have confirmed that this works as discussed here for v2.  We should have a prerelease version deployed to nuget sometime by the end of this week hopefully for you to try.  I will close this issue now but please feel free to open a new issue and/or reopen this one if you experience any problems with it.

**Comment by Mike-EEE** [comment]
FWIW, pre-release deployment for this feature can be found at these feeds:
https://www.myget.org/feed/wojtpl2/package/nuget/ExtendedXmlSerializer
https://www.nuget.org/packages/ExtendedXmlSerializer/2.0.0-alpha
